src/pgen/torus.cpp

```diff
@@ -295,21 +257,143 @@ void ProblemModifier(ParameterInput *pin) {
+  }
+}
+
+KOKKOS_FUNCTION
+Real lfish_calc(Real r, Real a) {
+  return (
+      ((std::pow(a, 2) - 2. * a * std::sqrt(r) + std::pow(r, 2)) *
+       ((-2. * a * r * (std::pow(a, 2) - 2. * a * std::sqrt(r) + std::pow(r, 2))) /
+            std::sqrt(2. * a * std::sqrt(r) + (-3. + r) * r) +
+        ((a + (-2. + r) * std::sqrt(r)) * (std::pow(r, 3) + std::pow(a, 2) * (2. + r))) /
+            std::sqrt(1 + (2. * a) / std::pow(r, 1.5) - 3. / r))) /
+      (std::pow(r, 3) * std::sqrt(2. * a * std::sqrt(r) + (-3. + r) * r) *
+       (std::pow(a, 2) + (-2. + r) * r)));
+}
+
+KOKKOS_FUNCTION
+Real log_enthalpy(const Real r, const Real th, const Real a, const Real rin, const Real l,
+                  Real &uphi) {
+  const Real sth = sin(th);
+  const Real cth = cos(th);
+
+  const Real DD = r * r - 2. * r + a * a;
+  const Real AA = (r * r + a * a) * (r * r + a * a) - DD * a * a * sth * sth;
+  const Real SS = r * r + a * a * cth * cth;
+
+  const Real thin = M_PI / 2.;
+  const Real sthin = sin(thin);
+  const Real cthin = cos(thin);
+
+  const Real DDin = rin * rin - 2. * rin + a * a;
+  const Real AAin =
+      (rin * rin + a * a) * (rin * rin + a * a) - DDin * a * a * sthin * sthin;
+  const Real SSin = rin * rin + a * a * cthin * cthin;
+  uphi = 0.0;
+  const Real lnh =
+      0.5 * std::log((1. + std::sqrt(1. + 4. * (l * l * SS * SS) * DD /
+                                              (AA * AA * sth * sth))) /
+                     (SS * DD / AA)) -
+      0.5 * std::sqrt(1. + 4. * (l * l * SS * SS) * DD / (AA * AA * sth * sth)) -
+      2. * a * r * l / AA -
+      (0.5 * std::log((1. + std::sqrt(1. + 4. * (l * l * SSin * SSin) * DDin /
+                                               (AAin * AAin * sthin * sthin))) /
+                      (SSin * DDin / AAin)) -
+       0.5 * std::sqrt(1. + 4. * (l * l * SSin * SSin) * DDin /
+                                (AAin * AAin * sthin * sthin)) -
+       2. * a * rin * l / AAin);
+  if (lnh > 0.0) {
+    Real expm2chi = SS * SS * DD / (AA * AA * sth * sth);
+    Real up1 = std::sqrt((-1. + std::sqrt(1. + 4. * l * l * expm2chi)) / 2.);
+    uphi = 2. * a * r * std::sqrt(1. + up1 * up1) / std::sqrt(AA * SS * DD) +
+           std::sqrt(SS / AA) * up1 / sth;
+  }
+  return lnh;
+}
+
+KOKKOS_FUNCTION
+Real ucon_norm(Real ucon[4], Real gcov[4][4]) {
+  Real AA = gcov[0][0];
+  Real BB = 2. * (gcov[0][1] * ucon[1] + gcov[0][2] * ucon[2] + gcov[0][3] * ucon[3]);
+  Real CC = 1. + gcov[1][1] * ucon[1] * ucon[1] + gcov[2][2] * ucon[2] * ucon[2] +
+            gcov[3][3] * ucon[3] * ucon[3] +
+            2. * (gcov[1][2] * ucon[1] * ucon[2] + gcov[1][3] * ucon[1] * ucon[3] +
+                  gcov[2][3] * ucon[2] * ucon[3]);
+  Real discr = BB * BB - 4. * AA * CC;
+  if (discr < 0) printf("discr = %g   %g %g %g\n", discr, AA, BB, CC);
+  PARTHENON_REQUIRE(discr >= 0, "discr < 0");
+  return (-BB - std::sqrt(discr)) / (2. * AA);
+}
+
+// TODO(JMM): Should this be elsewhere in Phoebus?
```

Comment:
I think this is the best spot for this right now, since it calculates global quantities (I'm not sure that `reduction::Min` and `reduction::Max` are going to behave outside of problem initilalization)

Reply:
:+1: